layers.go

```diff
@@ -1480,3 +1485,87 @@ func (r *layerStore) TouchedSince(when time.Time) bool {
 func (r *layerStore) Locked() bool {
 	return r.lockfile.Locked()
 }
+
+func newAdditionalROLayerStore(rundir string, layerdir string, driver drivers.Driver) (ROLayerStore, error) {
+	lockfile, err := GetLockfile(filepath.Join(layerdir, "layers.lock"))
+	if err != nil {
+		return nil, err
+	}
+	l := &additionalLayerStore{
+		layerStore: &layerStore{
+			lockfile:       lockfile, // use r/w lock for updating this store internally.
+			mountsLockfile: nil,
+			driver:         driver,
+			layerdir:       layerdir,
+			readOnly:       true, // configure this store as "read only" mode externally
+		},
+	}
+	if err := l.Load(); err != nil {
+		return nil, err
+	}
+	return l, nil
+}
+
+type additionalLayerStore struct {
+	*layerStore
+}
+
+func (r *additionalLayerStore) Load() error {
+	var layers []*Layer
+	byid := make(map[string]*Layer)
+	var idlist []string
+	files, err := ioutil.ReadDir(r.layerdir)
+	if err != nil {
+		return err
+	}
+	for _, f := range files {
+		if f.Name() == ".pool" || f.Name() == "layers.lock" {
+			continue
+		}
+		// TODO: check if the layer is healthy
+		l := &Layer{
+			ID:              f.Name(),
+			Created:         f.ModTime(),
+			CompressionType: archive.Uncompressed,
+			ReadOnly:        true,
+		}
+		layers = append(layers, l)
+		byid[f.Name()] = l
+		idlist = append(idlist, f.Name())
+	}
+	r.layers = layers
+	r.byid = byid
+	r.idindex = truncindex.NewTruncIndex(idlist)
+	return nil
+}
+
+func (r *additionalLayerStore) searchLayer(id string, annotations map[string]string) (*Layer, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	targetPrefix := "containers/image/target."
+	var kvs []string
+	for k, v := range annotations {
+		if strings.HasPrefix(k, targetPrefix) {
+			name := base64.StdEncoding.EncodeToString(
+				[]byte(strings.TrimPrefix(k, targetPrefix) + "=" + v))
+			if strings.HasPrefix(name, "rootfs.") {
+				return nil, fmt.Errorf("annotation must not have \"rootfs.\" prefix")
+			}
+			kvs = append(kvs, name)
+		}
+	}
+	target := filepath.Join(append([]string{r.layerdir, ".pool"}, append(kvs, "rootfs."+id)...)...)
```

Comment:
`annotations` is a map, which makes its iteration order unspecified, making the `target` value unstable. Besides, making this work depend on order of annotations in a manifest is unnecessarily brittle.

Reply:
Also, accessing the filesystem _at all_ if any of the required information is missing worries me conceptually. Given the base64, it’s _probably_ not a security risk, but why toy with fire?

Reply:
> Also, accessing the filesystem at all if any of the required information is missing worries me conceptually.

Cloud you elaborate you concern? `accessing the filesystem if any of the required information is missing` should simply fail and fallback to the normal pull procedure.

Reply:
The container image is, in general, not trusted by the consumer, i.e. it is a possible way to attack the container runtime.

Thus, every image metadata field and feature has to be evaluated as a possible way for a malicious attacker to manipulate the container runtime; given the attacker too much flexibility is always a bit of a risk (that has to be somehow weighted against utility of features, sure).

---

Having read a bit more about the design, I think the core issue that the reference + layer digest that are currently passed via `WithAnnotations` should not be a part of the _image format_ at all! Gathering this data and passing it through to `searchLayer` is an internal implementation detail of the “pull” process, entirely local to the pull operation; there is AFAICS no reason for that a data to show up in manifests uploaded to registries.

It seems to me that the relevant data is available, or can be collected, by `c/image/copy` before calling `TryReusingBlob`, and that should happen by just adding more well-specified parameters to (some future variant of) `TryReusingBlob`, not by adding an opaque `map[string]string` data transfer mechanism just because that avoids new parameters, and especially not by using a data transfer mechanism that results in long-term externally-observable manifest changes.

(Sure, that would imply incompatible changes to `types.ImageDestination`, which is a hassle and requires rather more process. That’s common to all the recently proposed image reuse / on-demand loading proposals for c/image, and while I appreciate that having to deal with that on top of the complexity of actually implementing the image reuse / on-demand loading is an extra obstacle, it is something that just needs to happen for the c/image feature set to grow — and it will happen sooner or later.)

Reply:
> It seems to me that the relevant data is available, or can be collected, by c/image/copy before calling TryReusingBlob, and that should happen by just adding more well-specified parameters to (some future variant of) TryReusingBlob, not by adding an opaque map[string]string data transfer mechanism just because that avoids new parameters, and especially not by using a data transfer mechanism that results in long-term externally-observable manifest changes.

Fixed this.
Now `*c/image.storageImageDestination.TryReusingBlobWithRef()` recieves reference through its parameter, not through anontations.